### PR TITLE
Block scoped classes

### LINKED ISSUE #2595: Class declarations should be renamed (like let/const) when they conflict with something in scope.
Class declarations are block scoped.  However, we emit them using 'var' declarations.  This will a same-named class in a nested scope to stomp on the outer class.  

I've looked at the renaming code in the emitter, and it's just complicated enough that i don't feel confident making the change here to fix things.  I'd recommend @JsonFreeman or @vladima to look at this.

### PR DESCRIPTION
**Pull request by vladima**
Fixes #2595

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
:+1:

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11749,23 +11749,23 @@ module ts {
...
                 getNodeLinks(n).resolvedSymbol ||
                 resolveName(n, n.text, SymbolFlags.Value | SymbolFlags.Alias, /*nodeNotFoundMessage*/ undefined, /*nameArg*/ undefined);
 
-            let isLetOrConst =
+            let hasMatchingMeaning =
                 symbol &&
-                (symbol.flags & SymbolFlags.BlockScopedVariable) &&
-                symbol.valueDeclaration.parent.kind !== SyntaxKind.CatchClause;
+                (symbol.flags & meaning) &&
+                (!(meaning & SymbolFlags.BlockScopedVariable) || symbol.valueDeclaration.parent.kind !== SyntaxKind.CatchClause);
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I don't think the check for `meaning & SymbolFlags.BlockScopedVariable` is necessary. The only way for the parent to be a catch clause is if it is also a block scoped variable.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -2777,34 +2776,61 @@ var __param = this.__param || function(index, decorator) { return function (targ
                 return getCombinedNodeFlags(node.parent);
             }
 
-            function renameNonTopLevelLetAndConst(node: Node): void {
+            function renameBlockScopedValueIfNecessary(node: Node, meaning: SymbolFlags): void {
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: I wouldn't call this meaning. I would call it symbolKind or something like that.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -11749,23 +11749,23 @@ module ts {
             return !!resolveName(location, name, SymbolFlags.Value, /*nodeNotFoundMessage*/ undefined, /*nameArg*/ undefined);
         }
 
-        function getBlockScopedVariableId(n: Identifier): number {
+        function getNamedValueId(n: Identifier, meaning: SymbolFlags): number {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Why is it necessary to take the specific meaning? Why can't this function simply always check for `SymbolFlags.BlockScopedVariable | SymbolFlags.Class`?

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -2777,34 +2776,61 @@ var __param = this.__param || function(index, decorator) { return function (targ
...
+                        return;
+                    }
+
+                    let list = getAncestor(node, SyntaxKind.VariableDeclarationList);
+                    if (list.parent.kind === SyntaxKind.VariableStatement) {
+                        container = list.parent.parent;
+                    }
+                }
+                else {
+                    if (node.parent.kind !== SyntaxKind.ClassDeclaration) {
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Would it be simpler to just include Class in BlockScopedVariable? We may end up needing to do this anyway for deep class declarations.

Better yet, maybe have another symbol flag called BlockScoped that is the union of Class and BlockScopedVariable.

File: src/compiler/emitter.ts (hunk 2 above)
Comment (vladima) [inline src/compiler/emitter.ts]: I will rename it to `expectedMeaning` - the intent behind this parameter is to denote the expected meaning for the name defined by `node`

File: src/compiler/types.ts [hunk 5]
```diff
@@ -1254,7 +1254,7 @@ module ts {
         // Returns the constant value this property access resolves to, or 'undefined' for a non-constant
         getConstantValue(node: EnumMember | PropertyAccessExpression | ElementAccessExpression): number;
         resolvesToSomeValue(location: Node, name: string): boolean;
-        getBlockScopedVariableId(node: Identifier): number;
+        getBlockScopedValueId(n: Identifier): number;
```
Comment (JsonFreeman) [inline src/compiler/types.ts]: It has now switched from `node` to `n`. Please rename it back

File: src/compiler/emitter.ts [hunk 6]
```diff
@@ -120,7 +119,7 @@ var __param = this.__param || function(index, decorator) { return function (targ
 
             let generatedNameSet: Map<string> = {};
             let nodeToGeneratedName: string[] = [];
-            let blockScopedVariableToGeneratedName: string[];
+            let blockScopedNameToGeneratedName: string[];
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: In the checker these are called blockScopedValues but here they are blockScopedNames. Is there a reason for the terminology difference?

File: src/compiler/emitter.ts (hunk 4 above)
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: I'm trying to think when you would have a class symbol where the node parent is not a class declaration. I guess a reverse clodule. Better to be defensive I suppose

**Comment by JsonFreeman** [comment]
:+1:

**Comment by mhegazy** [comment]
:+1:

**Comment by vladima** [comment]
this PR is subsumed by #3380

**Pull request closed without merging**